day3/slice_time_image.py

```diff
@@ -46,11 +56,37 @@ def slice_time_image(img, slice_times, TR):
     st_img : Nifti1Image
         A new copy of the input image with slice-time interpolation applied
     """
+
     # Get the image data as an array;
+    data = img.get_data()       # The current function structure leads to many loadings of the image file, making the process slower and increasing memory demands
+    n_x_voxels = data.shape[0]    
+    n_y_voxels = data.shape[1]    
+    n_slices = data.shape[2] # number of slices on z dimension
+    n_tr = data.shape[3]
+    
     # Make a new empty array "interp_data" to hold the interpolated data;
+    interp_data = np.zeros(data.shape)
+
     # Do the interpolation;
+    tr_onsets = np.arange(n_tr) * TR  # define onset of each TR in sec
+
+    # for each slice interpolate time series to the onset of the first slice acquired in the TR
+    for iz in range(n_slices):
+        all_iz_times = tr_onsets + slice_times[iz]
+        # for each voxel in the slice do an interpolation; could possibly reshape matrix first, but more difficult to read and then have to inverse reshape
+        for ix in range(n_x_voxels):
+            for iy in range (n_y_voxels):
+                voxel_timeseries = data[ix,iy,iz,:]
+                interp = InterpolatedUnivariateSpline(all_iz_times, voxel_timeseries, k=1)
+                interp_data[ix,iy,iz,:] = interp(tr_onsets)  #we could potentially also interpolate for all n_slices intervals and get n_tr * n_slices time points.
+
+    # debug
```

Comment:
Generally remove commented out stuff when you've done, it can be distracting for the reviewer.